### PR TITLE
add scoring strategy for mark dups in Spark. Thus fixes 1 integration…

### PR DESCRIPTION
**Pull request by akiezun**
… test.

partial fix for #1042 - reenabled testStackOverFlowPairSetSwap - the failure was due to scoring strategy using by picard (total ref bases) vs spark (sum of quals). Spark did not even have a pluggable scoring strategy. Now it does and the test passes.

For @davidadamsphd

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java [hunk 1]
```diff
@@ -68,6 +69,9 @@
     @Argument(doc = "the join strategy for reference bases and known variants", shortName = "joinStrategy", fullName = "joinStrategy", optional = true)
     private JoinStrategy joinStrategy = JoinStrategy.BROADCAST;
 
+    @Argument(shortName = "DS", doc = "The scoring strategy for choosing the non-duplicate among candidates.")
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java]: Please also add a fullName.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java [hunk 2]
```diff
@@ -38,6 +39,9 @@
             shortName = "M", fullName = "METRICS_FILE")
     protected String metricsFile;
 
+    @Argument(shortName = "DS", doc = "The scoring strategy for choosing the non-duplicate among candidates.")
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java]: Please add a fullName.

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/MarkDuplicatesScoringStrategy.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    public int score(final GATKRead read){
+        Utils.nonNull(read);
+        return scoring.applyAsInt(read);
+    }
+
+    //Bases below this quality will not be included in picking the best read from a set of duplicates.
+    public static final int MIN_BASE_QUAL = 15;
+
+    /**
+     * How to assign a score to the read in MarkDuplicates (so that we pick the best one to be the non-duplicate).
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/MarkDuplicatesScoringStrategy.java]: Please update this comment to be more specific to the function.

**Comment by davidadamsphd** [comment]
I don't see any serious issues, but I have broader question: under what cases would use we one scoring strategy over another?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/MarkDuplicatesScoringStrategy.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/MarkDuplicatesScoringStrategy.java]: done

**Comment by akiezun** [comment]
i'm not sure but TOTAL_MAPPED_REFERENCE_LENGTH is the default in picard and the default used in production.

**Pull request merged**